syntax = "proto3";

//Represents an empty message, for RPC calls that don't take any arguments
message Empty {}

//Represents a three dimensional vector
message Vector3D
{
	float x = 1;
	float y = 2;
	float z = 3;
}

message InstanceIndex
{
	int32 index = 1;
}

//Represents a transform in 3D space
message Transform
{
	Vector3D position = 1;
	Vector3D rotation = 2;
	Vector3D scale = 3;
}

message InstanceTransform
{
	int32 index = 1;
	Transform transform = 2;
}

message InstanceForce
{
	int32 index = 1;
	Vector3D force = 2;
}

service ActorOperation
{
	//Sets the transform of the actor
	rpc SetActorTransform (InstanceTransform) returns (Empty) {}
	
	//Retrieves the current transform of the actor
	rpc GetActorTransform (InstanceIndex) returns (Transform) {}
	
	//Applies force to the actor
	rpc ApplyForceToActor (InstanceForce) returns (Empty) {}
}
